### PR TITLE
fix(observability): fix write to db isssue for observability

### PR DESCRIPTION
**Pull request by rajaSahil**
**Purpose of PR?**:

- Fix concurrent write issue which was happening because of using lock in a different goroutine. 
- Fixes write to db for karmor summary

*NOTE*
- There is another issue with protobuf which is causing DE to panic because of invalid go types.

**Checklist:**
- [x] Bug fix. Fixes #<issue number>
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: src/observability/publisher.go [hunk 1]
```diff
@@ -20,19 +20,19 @@ func ProcessSystemSummary() {
 	tempSummarizerMap := common.MoveMap(SummarizerMap)
 	SummarizerMapMutex.Unlock()
 
-	var ProcessSystemSummaryWg sync.WaitGroup
+	aggregateSummaryMap(tempSummarizerMap)
 
 	if cfg.GetCfgObservabilityWriteToDB() {
+		var ProcessSystemSummaryWg sync.WaitGroup
```
Comment (seswarrajan) [inline src/observability/publisher.go:26]:
```suggestion
		ProcessSystemSummaryWg := sync.WaitGroup{}
```

**Review by seswarrajan — CHANGES_REQUESTED** [review]

**Review by achrefbensaad — APPROVED** [review]
LGTM

File: src/observability/publisher.go [hunk 2]
```diff
@@ -20,19 +20,19 @@ func ProcessSystemSummary() {
...
 		ProcessSystemSummaryWg.Add(1)
 		go UpsertSummaryCronJob(tempSummarizerMap, &ProcessSystemSummaryWg)
+		ProcessSystemSummaryWg.Wait()
```
Comment (stefin9898) [inline src/observability/publisher.go:29]: We can remove the entire ProcessSystemSummaryWg by removing the call to `UpsertSummaryCronJob` as a goroutine

Comment (rajaSahil) [reply src/observability/publisher.go:29]: I think it will have some performance implications, let's use go routines for this.

**Comment by seswarrajan** [comment]
One more suggestion. Instead of using wg wait and sync, why not write to DB one by one, in line 50, after Publish. We do have a common handle for DB and we are not going to open a new handle every time while we write to DB. 

Will that work ?

Since the logs/summary data is aggreagted, we will have less data and writing to DB in loop individually will not cost much. Please correct me if i am wrong.

File: src/observability/publisher.go (hunk 2 above)
Comment (stefin9898) [reply src/observability/publisher.go:29]: Since it is waiting at `ProcessSystemSummaryWg.Wait()` then why should there be a goroutine?

**Comment by stefin9898** [comment]
Yes @seswarrajan that make sense! We can just iterate the tempSummarizerMap once and publish / write to db if the flags are enabled for the same

File: src/observability/publisher.go (hunk 2 above)
Comment (rajaSahil) [reply src/observability/publisher.go:29]: What I  was thinking of is to spawn two go routines for both the task, but I haven't tried to implement it. I will be trying it, so let's have it for now .

**Comment by rajaSahil** [comment]
> One more suggestion. Instead of using wg wait and sync, why not write to DB one by one, in line 50, after Publish. We do have a common handle for DB and we are not going to open a new handle every time while we write to DB.
> 
> Will that work ?
> 
> Since the logs/summary data is aggreagted, we will have less data and writing to DB in loop individually will not cost much. Please correct me if i am wrong.

No, having seperate configs make sense to me!

**Comment by stefin9898** [comment]
> No, having seperate configs make sense to me!

Yes, it will be having separate configs. I believe @seswarrajan meant to iterate the  tempSummarizerMap once. right?

**Comment by seswarrajan** [comment]
> > One more suggestion. Instead of using wg wait and sync, why not write to DB one by one, in line 50, after Publish. We do have a common handle for DB and we are not going to open a new handle every time while we write to DB.
> > Will that work ?
> > Since the logs/summary data is aggreagted, we will have less data and writing to DB in loop individually will not cost much. Please correct me if i am wrong.
> 
> No, having seperate configs make sense to me!

@rajaSahil The config still remains untouched. We will have separate configs. But, we need to call write to db after Publish. For each log. That write will fall inside the condition check.

**Comment by rajaSahil** [comment]
@seswarrajan Unable to get you. Why not we spawn two different goroutines and have it processed at the same time?

File: src/observability/publisher.go [hunk 3]
```diff
@@ -20,19 +20,19 @@ func ProcessSystemSummary() {
 	tempSummarizerMap := common.MoveMap(SummarizerMap)
 	SummarizerMapMutex.Unlock()
 
-	var ProcessSystemSummaryWg sync.WaitGroup
+	aggregateSummaryMap(tempSummarizerMap)
```
Comment (vishnusomank) [inline src/observability/publisher.go:23]: Do we really need to aggregate and store in the DB? IMO we should keep the raw data in DB and use agg before publishing it to SaaS? WDYT @rajaSahil @seswarrajan

Comment (rajaSahil) [reply src/observability/publisher.go:23]: We are anyway aggregating  the data before pushing to SAAS, why can't we store aggregated result in the DB? Is it going to create any issues? We anyway aggregate all the file data for karmor summary, so we can do 1st level of aggregation over here which can increase performance for karmor summary also.

Comment (vishnusomank) [reply src/observability/publisher.go:23]: It is not going to create any issues. And the agg is false by default in karmor summary right? That is why I suggested to keep it raw.

**Review by Ankurk99 — APPROVED** [review]
LGTM!

**Review by seswarrajan — APPROVED** [review]

**Pull request merged**